File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ResyncLanguageNamesTask.java
package org.wycliffeassociates.translationrecorder.ProjectManager.tasks.resync;

import android.content.Context;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONException;
import org.wycliffeassociates.translationrecorder.database.ProjectDatabaseHelper;
import org.wycliffeassociates.translationrecorder.project.components.Language;
import org.wycliffeassociates.translationrecorder.project.ParseJSON;
import org.wycliffeassociates.translationrecorder.utilities.Task;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.FileNotFoundException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

/**
 * Created by sarabiaj on 12/14/2016.
 */

public class ResyncLanguageNamesTask extends Task {
    Context mCtx;
    ProjectDatabaseHelper db;
    boolean isLocal = false;
    Uri localFile;
    Handler handler;

    public ResyncLanguageNamesTask(int taskTag, Context ctx, ProjectDatabaseHelper db) {
        super(taskTag);
        mCtx = ctx;
        this.db = db;
        this.handler = new Handler(Looper.getMainLooper());
    }

    public ResyncLanguageNamesTask(int taskTag, Context ctx, ProjectDatabaseHelper db, Uri uri) {
        this(taskTag, ctx, db);
        this.isLocal = true;
        this.localFile = uri;
    }

    @Override
    public void run() {
        String json = isLocal ? loadJsonFromFile() : loadJsonFromUrl();
        try {
            JSONArray jsonObject = new JSONArray(json);
            ParseJSON parseJSON = new ParseJSON(mCtx);
            Language[] languages = parseJSON.pullLangNames(jsonObject);
            db.addLanguages(languages);
            onTaskCompleteDelegator();
            handler.post(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(mCtx, "Languages successfully updated!", Toast.LENGTH_SHORT).show();
                }
            });
        } catch (final JSONException e) {
            e.printStackTrace();
            onTaskErrorDelegator();
            handler.post(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(mCtx, "Invalid json file", Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

    private String loadJsonFromUrl() {
        try {
            URL url = new URL("http://td.unfoldingword.org/exports/langnames.json");
            HttpURLConnection urlConnection = null;

            try {
                urlConnection = (HttpURLConnection) url.openConnection();
                InputStream in = new BufferedInputStream(urlConnection.getInputStream());
                BufferedReader reader = new BufferedReader(new InputStreamReader(in));
                StringBuilder json = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    json.append(line);
                }
                reader.close();
                return json.toString();
            } catch (final IOException e) {
                e.printStackTrace();
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(mCtx, e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
                return "";
            } finally {
                urlConnection.disconnect();
            }
        } catch (MalformedURLException e) {
            e.printStackTrace();
            return "";
        }
    }

    private String loadJsonFromFile() {
        try (InputStream is = mCtx.getContentResolver().openInputStream(localFile)) {
            BufferedReader reader = new BufferedReader(new InputStreamReader(is));
            StringBuilder json = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                json.append(line);
            }
            reader.close();
            return json.toString();
        } catch (final FileNotFoundException e) {
            e.printStackTrace();
            handler.post(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(mCtx, e.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
            return "";
        } catch (IOException e) {
            e.printStackTrace();
            onTaskErrorDelegator();
            return "";
        }
    }
}

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ResyncLanguageNamesTask.kt
package org.wycliffeassociates.translationrecorder.ProjectManager.tasks.resync

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import org.json.JSONArray
import org.json.JSONException
import org.wycliffeassociates.translationrecorder.R
import org.wycliffeassociates.translationrecorder.database.IProjectDatabaseHelper
import org.wycliffeassociates.translationrecorder.persistance.AssetsProvider
import org.wycliffeassociates.translationrecorder.project.ParseJSON
import org.wycliffeassociates.translationrecorder.project.components.Language
import org.wycliffeassociates.translationrecorder.utilities.Task
import java.io.BufferedInputStream
import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

/**
 * Created by sarabiaj on 12/14/2016.
 */
class ResyncLanguageNamesTask private constructor(
    taskTag: Int,
    private val mCtx: Context,
    private val db: IProjectDatabaseHelper,
    private val assetsProvider: AssetsProvider,
    private val context: Context
) : Task(taskTag) {
    private var isLocal: Boolean = false
    private var localFile: Uri? = null
    private var remoteUrl: String? = null
    private val handler: Handler = Handler(Looper.getMainLooper())

    constructor(
        taskTag: Int,
        ctx: Context,
        db: IProjectDatabaseHelper,
        assetsProvider: AssetsProvider,
        context: Context,
        url: String?
    ) : this(taskTag, ctx, db, assetsProvider, context) {
        this.remoteUrl = url
    }

    constructor(
        taskTag: Int,
        ctx: Context,
        db: IProjectDatabaseHelper,
        assetsProvider: AssetsProvider,
        context: Context,
        uri: Uri?
    ) : this(
        taskTag,
        ctx,
        db,
        assetsProvider,
        context
    ) {
        this.isLocal = true
        this.localFile = uri
    }

    override fun run() {
        val json = if (isLocal) loadJsonFromFile() else loadJsonFromUrl()
        try {
            val jsonObject = JSONArray(json)
            val parseJSON = ParseJSON(assetsProvider, context)
            val languages: List<Language> = parseJSON.pullLangNames(jsonObject)
            db.addLanguages(languages)
            onTaskCompleteDelegator()
            handler.post {
                Toast.makeText(
                    mCtx,
                    mCtx.getString(R.string.languages_updated),
                    Toast.LENGTH_SHORT
                ).show()
            }
        } catch (e: JSONException) {
            e.printStackTrace()
            onTaskErrorDelegator()
            handler.post {
                Toast.makeText(
                    mCtx,
                    mCtx.getString(R.string.invalid_json),
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }

    private fun loadJsonFromUrl(): String {
        try {
            val url = URL(this.remoteUrl)
            var urlConnection: HttpURLConnection? = null

            try {
                urlConnection = url.openConnection() as HttpURLConnection
                val `in`: InputStream = BufferedInputStream(urlConnection.inputStream)
                val reader = BufferedReader(InputStreamReader(`in`))
                val json = StringBuilder()
                var line: String?
                while ((reader.readLine().also { line = it }) != null) {
                    json.append(line)
                }
                reader.close()
                return json.toString()
            } catch (e: IOException) {
                e.printStackTrace()
                handler.post { Toast.makeText(mCtx, e.message, Toast.LENGTH_SHORT).show() }
                return ""
            } finally {
                urlConnection!!.disconnect()
            }
        } catch (e: MalformedURLException) {
            e.printStackTrace()
            return ""
        }
    }

    private fun loadJsonFromFile(): String {
        try {
            mCtx.contentResolver.openInputStream(localFile!!).use { `is` ->
                val reader = BufferedReader(InputStreamReader(`is`))
                val json = StringBuilder()
                var line: String?
                while ((reader.readLine().also { line = it }) != null) {
                    json.append(line)
                }
                reader.close()
                return json.toString()
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
            handler.post { Toast.makeText(mCtx, e.message, Toast.LENGTH_SHORT).show() }
            return ""
        } catch (e: IOException) {
            e.printStackTrace()
            onTaskErrorDelegator()
            return ""
        }
    }
}
